examples/5_secret_storage.py

```diff
@@ -0,0 +1,419 @@
+# ###################################
+
+import asyncio
+import hmac
+import json
+import logging
+import secrets
+import sys
+from base64 import b64decode, b64encode
+from typing import Any, Dict, List, Union
+
+from pykeybasebot import Bot, KVStoreClient
+from pykeybasebot.errors import DeleteNonExistentError, RevisionError
+from pykeybasebot.types import keybase1
+
+logging.basicConfig(level=logging.DEBUG)
+
+if "win32" in sys.platform:
+    # Windows specific event-loop policy
+    asyncio.set_event_loop_policy(
+        asyncio.WindowsProactorEventLoopPolicy()  # type: ignore
+    )
+
+
+class CachedBot(Bot):
+    """
+        Custom bot that has some stateful clients.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.secret_kvstore_client = SecretKeyKVStoreClient(self)
+        self.cached_secret_kvstore_client = CachedKVStoreClient(
+            self, self.secret_kvstore_client
+        )
+        super().__init__(*args, **kwargs)
+
+    @property
+    def cached_secret_kvstore(self):
+        return self.cached_secret_kvstore_client
+
+
+class CachedKVStoreClient:
+    """
+    CachedKVStoreClient uses a cache to keep track of the most recently fetched value and
+    revision for each key. To handle concurrent updates, it attempts to update with
+    the most recently fetched revision + 1; if it fails, it does a "get" and updates
+    the cache, and returns that "get" result.
+    """
+
+    def __init__(self, bot, client):
+        # self.cache = {entryKey: {"revision": int, entryValue: {} or None}}
+        self.cache: Dict[Any, Any] = {}
+        self.kvstore = client
+
+    # note that we expect entryValues to be JSON objects
+    def update_cache(
+        self, entry_key: str, entry_value: Union[None, Dict[str, str]], revision: int
+    ):
+        self.cache[entry_key] = {"info": entry_value, "revision": revision}
+
+    # returns a copy of the cached value for a given entry_key
+    def get_cached(self, entry_key: str):
+        cached = self.cache[entry_key].copy() if entry_key in self.cache else None
+        if cached is not None:
+            cached["info"] = (
+                cached["info"].copy() if cached["info"] is not None else None
+            )
+        return cached
+
+    async def put(
+        self,
+        team: str,
+        namespace: str,
+        entry_key: str,
+        entry_value: Dict[str, str],
+        revision: Union[int, None] = None,
+    ) -> Union[keybase1.KVPutResult, keybase1.KVGetResult]:
+        try:
+            res: keybase1.KVPutResult = await self.kvstore.put(
+                team, namespace, entry_key, entry_value, revision
+            )
+            self.update_cache(entry_key, entry_value, res.revision)
+            return res  # successful put. return KVPUtResult
+        except RevisionError:
+            # refresh cached value
+            curr_info = await self.get(team, namespace, entry_key)
+            return curr_info  # failed put. return KVGetResult.
```

Comment:
not sure i love this pattern. a PUT could return the PUT result _OR_ a GET result?
not a blocker. i just think this is a little bit of a weird API.